stackrox/stackrox
# ROX-19234: Add conditional RBAC for Vulnerabilities Reports 2.0

## Description

Set precedent to release even XL features with conditional rendering according to role-based access control.

Make only changes that are testable so not to invalidate testing of the feature itself.

What cause me to change original decision to wait until 4.3 release?
* Obsolete Access requirement to clone, create, edit report.
* Lack of conditional rendering for reports table row actions and report page **Actions** button.

Adapt from Reports 1.0 in #7433

### Analysis

1. VulnReportingPage
    * `canReadWriteReports` includes obsolete Access resource for access scopes superseded by collections
    * Integration is required only for select options to write, because notifier integration name is included in report configuration response.
2. VulnReportsPage
    * Ditto the above
    * Table does not have conditional rendering for row actions.
3. ViewVulnReportsPage
    * Page does not have conditional rendering for **Action** button.
3. CollectionSelection:
    * ReportParametersForm uses permissions information only for `allowCreate` prop.
    * Collections modal is like a link to another container, therefore needs `isRouteEnabled` function call. In this case, collections depend not only on implicit WorkflowAdministation same as reports, but also on READ_ACCESS for Deployment to display collection results in Review step of wizard.
4. TechPreviewBanner unconditionally renders link to another container:
    * From Reports 2.0 to Reports 1.0 both routes have the same resource requirements.
    * From Workload CVEs to Vulnerability Management 1.0 the latter has more resource requirements.

### Residue

1. Factor out `hasWriteAccessForReport` criterion.
2. Evaluate pro and con to add for conditional rendering of **Create email notifier** button an `isRouteEnabled('integrations')` function call, which is redundant with READ_ACCESS for Integration resource that is implied by `hasWriteAccessForReport` value.

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.css: 0 = 3761815 - 3761815
        total: 336 = 9969784 - 9969448
    * `ls -al build/static/js/*.js | wc -l`
        0 = 80 - 80 files
3. `yarn start` in ui

### Manual testing

1. Visit /main/vulnerabilities/reports with only READ_ACCESS for VulnerabilityAdministration resource.

    * See preview banner **with** link but no such link in sidebar navigation.
        ![reports_READ_ACCESS_original](https://github.com/stackrox/stackrox/assets/11862657/2483d668-bf6e-45a6-8af8-e5a91ed6faa0)

    * See preview banner **without** link because route is not enabled.
        ![reports_READ_ACCESS_TechPreviewBanner](https://github.com/stackrox/stackrox/assets/11862657/ab1e7cc4-81a5-4ba2-a79c-3faa4d4a394b)

    * See absence of row actions.
        ![reports_READ_ACCESS_without_ActionsColumn](https://github.com/stackrox/stackrox/assets/11862657/667d4897-ebfc-4c23-b1c8-fb191f1c9890)

    * Click a link to open a report page and see absence of **Actions** button. 
        ![report_READ_ACCESS_without_Actions](https://github.com/stackrox/stackrox/assets/11862657/9b955745-e0cc-4c2d-8f9b-740da9e82d97)

2. Visit /main/vulnerabilities/reports with READ_WRITE_ACCESS for VulnerabilityAdministration also READ_ACCESS for Image and Integration resources.

    * See **Create report** button and row actions. No picture of original behavior.

3. Click **Edit report** in row actions to see report wizard.

    * See absence of **Create collection** button without READ_ACCESS for Deployment therefore route is not enabled.
        ![report_without_Create_collection](https://github.com/stackrox/stackrox/assets/11862657/fb3a0711-3fae-451b-a5ec-d4d50205fd50)

    * See presence of **Create collection** button with READ_ACCESS for Deployment therefore route is enabled.
        ![report_with_Create_collection](https://github.com/stackrox/stackrox/assets/11862657/b4de610c-42ac-43f3-af94-b62b33ff07dc)

    * See absence of **Create email notifier** button with READ_ACCESS for Integration.
        ![report_without_Create_email_notifier](https://github.com/stackrox/stackrox/assets/11862657/d011e3bd-c501-4642-bbc1-9df20e3b5b72)

    * See presence of **Create email notifier** button with READ_WRITE_ACCESS for Integration.
        ![report_with_Create_email_notifier](https://github.com/stackrox/stackrox/assets/11862657/972e7e20-d8de-4be6-8aa2-a33a4cadbea5)
